src/sinks/splunk_hec/common/service.rs

```diff
@@ -74,13 +150,265 @@ pub struct HttpRequestBuilder {
+            .await;
+
+        mock_server
+    }
+
+    #[tokio::test]
+    async fn acknowledgements_enabled_on_server() {
+        let ack_response = |req: &Request| {
+            let req = serde_json::from_slice::<HecAckStatusRequest>(req.body.as_slice()).unwrap();
+            ResponseTemplate::new(200).set_body_json(HecAckStatusResponse {
+                acks: req
+                    .acks
+                    .into_iter()
+                    .map(|ack_id| (ack_id, true))
+                    .collect::<HashMap<_, _>>(),
+            })
+        };
+        let mock_server = get_hec_mock_server(true, ack_response).await;
+
+        let acknowledgements_config = HecClientAcknowledgementsConfig {
+            query_interval: NonZeroU8::new(1).unwrap(),
+            ..HecClientAcknowledgementsConfig::default()
+        };
+        let mut service = get_hec_service(mock_server.uri(), acknowledgements_config);
+
+        let mut responses = (0..10)
+            .map(|_| service.call(get_hec_request()))
+            .collect::<FuturesUnordered<_>>();
+        while let Some(response) = responses.next().await {
+            assert_eq!(EventStatus::Delivered, response.unwrap().event_status)
+        }
+    }
+
+    #[tokio::test]
+    async fn acknowledgements_disabled_on_server() {
+        let ack_response = |_: &Request| ResponseTemplate::new(400);
+        let mock_server = get_hec_mock_server(false, ack_response).await;
+
+        let acknowledgements_config = HecClientAcknowledgementsConfig {
+            query_interval: NonZeroU8::new(1).unwrap(),
+            ..HecClientAcknowledgementsConfig::default()
+        };
+        let mut service = get_hec_service(mock_server.uri(), acknowledgements_config);
+
+        let request = get_hec_request();
+        let response = service.call(request).await.unwrap();
+        assert_eq!(EventStatus::Delivered, response.event_status)
+    }
+
+    #[tokio::test]
+    async fn acknowledgements_enabled_on_server_retry_limit_exceeded() {
+        let ack_response = |req: &Request| {
+            let req = serde_json::from_slice::<HecAckStatusRequest>(req.body.as_slice()).unwrap();
+            ResponseTemplate::new(200).set_body_json(HecAckStatusResponse {
+                acks: req
+                    .acks
+                    .into_iter()
+                    .map(|ack_id| (ack_id, false))
+                    .collect::<HashMap<_, _>>(),
+            })
+        };
+        let mock_server = get_hec_mock_server(true, ack_response).await;
+
+        let acknowledgements_config = HecClientAcknowledgementsConfig {
+            query_interval: NonZeroU8::new(1).unwrap(),
+            retry_limit: NonZeroU8::new(1).unwrap(),
+        };
+        let mut service = get_hec_service(mock_server.uri(), acknowledgements_config);
+
+        let request = get_hec_request();
+        let response = service.call(request).await.unwrap();
+        assert_eq!(EventStatus::Failed, response.event_status)
```

Comment:
Am I right in thinking this is the only way the server can signal rejection? If there is another way to signal rejection, then this might need another test.

Reply:
Right, currently the only way to receive `EventStatus::Failed` is for all of an `ack_id`'s retries to expire. In that case, the `tx` used to send back an `EventStatus` is dropped and we default to `Failed`.